src/Data/Nat/Primality/Factorisation.agda

```diff
@@ -0,0 +1,210 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Prime factorisation of natural numbers and its properties
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Data.Nat.Primality.Factorisation where
+
+open import Data.Empty using (⊥-elim)
+open import Data.Nat.Base
+open import Data.Nat.Divisibility using (_∣_; _∣?_; quotient; quotient∣n; ∣-trans; ∣1⇒≡1; divides)
+open import Data.Nat.Properties
+open import Data.Nat.Induction using (<-Rec; <-rec)
+open import Data.Nat.Primality using (Prime; euclidsLemma; ∣p⇒≡1∨≡p; Prime⇒NonZero)
+open import Data.Nat.Primality.Rough using (_Rough_; 2-rough-n; extend-∤; roughn∧∣n⇒prime)
+open import Data.Product as Π using (∃-syntax; _,_; proj₁; proj₂)
+open import Data.List.Base using (List; []; _∷_; _++_; product)
+open import Data.List.Membership.Propositional using (_∈_)
+open import Data.List.Membership.Propositional.Properties using (∈-∃++)
+open import Data.List.Relation.Unary.All as All using (All; []; _∷_)
+open import Data.List.Relation.Unary.Any using (here; there)
+open import Data.List.Relation.Binary.Permutation.Propositional as ↭
+  using (_↭_; prep; swap; ↭-refl; refl; module PermutationReasoning)
+open import Data.List.Relation.Binary.Permutation.Propositional.Properties using (product-↭; All-resp-↭; shift)
+open import Data.Sum.Base using (inj₁; inj₂)
+open import Function.Base using (_$_; _∘_; _|>_; flip)
+open import Relation.Nullary.Decidable using (yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+open import Relation.Binary.PropositionalEquality using (_≡_; refl; cong; module ≡-Reasoning)
+
+------------------------------------------------------------------------
+-- Core definition
+
+record PrimeFactorisation (n : ℕ) : Set where
+  field
+    factors : List ℕ
+    isFactorisation : product factors ≡ n
+    factorsPrime : All Prime factors
+
+open PrimeFactorisation public using (factors)
+open PrimeFactorisation
+
+------------------------------------------------------------------------
+-- Finding a factorisation
+
+-- this builds up three important things:
+-- * a proof that every number we've gotten to so far has increasingly higher
+--   possible least prime factor, so we don't have to repeat smaller factores
+--   over and over (this is the "k" and "k-rough-n" parameters)
+-- * a witness that this limit is getting closer to the number of interest, in a
+--   way that helps the termination checker (the k≤n parameter)
+-- * a proof that we can factorise any smaller number, which is useful when we
+--   encounter a factor, as we can then divide by that factor and continue from
+--   there without termination issues
+
+private
+  pattern 2+ n = suc (suc n)
+  pattern 2≤2+n = s≤s (s≤s z≤n)
+  pattern 1<2+n = 2≤2+n
```

Comment:
Apologies for insisting on having removed such things from earlier versions of the library. Do we need to reinstate them full-bloodedly in `Data.Nat.Base`?

Reply:
I think these were snuck in by @gallais here so I'm going to refer this to him